### PR TITLE
Add support for Hazelcast YAML configuration

### LINKED ISSUE #16074: Support YAML for Hazelcast configuration
Hazelcast has been traditionally configured with an XML file, `hazelcast.xml` for servers and `hazelcast-client.xml` for clients, and this is supported by Spring Boot Autoconfiguration.

Hazelcast is enhancing this to allow the same to be done with YAML, see https://docs.hazelcast.org/docs/latest-dev/manual/html-single/#configuring-declaratively-yaml

As an enhancement, Spring Boot should auto-configure a Hazelcast instance based on the presence of this YAML in the same way as is done for the presence of XML

### PR DESCRIPTION
**Pull request by leszko**
Since Hazlecast 3.12, YAML configuration format is supported in
addition to XML. Therefore, this change makes Spring Boot automatically
discover not only `hazelcast.xml` (and `hazelcast-client.xml`), but also
`hazelcast.yaml` (and `hazelcast-client.yaml`).

fix #16074 

Signed-off-by: Rafal Leszko <rafal@hazelcast.com>

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastClientFactory.java [hunk 1]
```diff
@@ -59,7 +60,14 @@ public HazelcastClientFactory(ClientConfig clientConfig) {
 	private ClientConfig getClientConfig(Resource clientConfigLocation)
 			throws IOException {
 		URL configUrl = clientConfigLocation.getURL();
-		return new XmlClientConfigBuilder(configUrl).build();
+		String configFileName = configUrl.getFile();
+
+		if (configFileName.endsWith(".xml")) {
+			return new XmlClientConfigBuilder(configUrl).build();
+		}
+		else { // yaml config available in the classpath
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastClientFactory.java]: What does this comment mean?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastInstanceFactory.java [hunk 2]
```diff
@@ -61,7 +62,16 @@ public HazelcastInstanceFactory(Config config) {
 
 	private Config getConfig(Resource configLocation) throws IOException {
 		URL configUrl = configLocation.getURL();
-		Config config = new XmlConfigBuilder(configUrl).build();
+		String configFileName = configUrl.getFile();
+
+		Config config;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastInstanceFactory.java]: Rather than having a field initialization, I'd create a private method that returns the `Config` instance.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastServerConfiguration.java [hunk 3]
```diff
@@ -77,7 +77,8 @@ public HazelcastInstance hazelcastInstance(Config config) {
 
 		ConfigAvailableCondition() {
 			super(CONFIG_SYSTEM_PROPERTY, "file:./hazelcast.xml",
-					"classpath:/hazelcast.xml");
+					"classpath:/hazelcast.xml", "file:./hazelcast.yaml",
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastServerConfiguration.java]: Same ordering remark

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfigurationTests.java [hunk 4]
```diff
@@ -39,6 +39,8 @@
 	@Test
 	public void defaultConfigFile() {
 		// no hazelcast-client.xml and hazelcast.xml is present in root classpath
+		// this also asserts that XML has priority over YAML
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfigurationTests.java]: It seems wrong to me that XML on the classpath has priority over YAML on the file system. Is this really what Hazelcast does?

File: spring-boot-project/spring-boot-autoconfigure/src/test/resources/hazelcast.xml [hunk 5]
```diff
@@ -5,7 +5,6 @@
 	<map name="defaultCache" />
 	<network>
 		<join>
-			<tcp-ip enabled="false" />
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/resources/hazelcast.xml]: This seems an unrelated change. I am happy to polish the support in a separate commit though. Can you explain why you've removed this line?

File: spring-boot-project/spring-boot-autoconfigure/src/test/resources/org/springframework/boot/autoconfigure/hazelcast/hazelcast-specific.xml [hunk 6]
```diff
@@ -2,16 +2,13 @@
 		   xmlns="http://www.hazelcast.com/schema/config"
 		   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
 
-	<queue name="foobar"/>
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/resources/org/springframework/boot/autoconfigure/hazelcast/hazelcast-specific.xml]: Same remark. Is this related to the main purpose of this PR?

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastInstanceFactory.java (hunk 2 above)
Comment (leszko) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastInstanceFactory.java]: Changed

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfigurationTests.java (hunk 4 above)
Comment (leszko) [reply spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastAutoConfigurationTests.java]: yeah, as explained in the other comment.

File: spring-boot-project/spring-boot-autoconfigure/src/test/resources/hazelcast.xml (hunk 5 above)
Comment (leszko) [reply spring-boot-project/spring-boot-autoconfigure/src/test/resources/hazelcast.xml]: Yeah, it's just not used. Reverted. Can be sent in a separate PR.

File: spring-boot-project/spring-boot-autoconfigure/src/test/resources/org/springframework/boot/autoconfigure/hazelcast/hazelcast-specific.xml (hunk 6 above)
Comment (leszko) [reply spring-boot-project/spring-boot-autoconfigure/src/test/resources/org/springframework/boot/autoconfigure/hazelcast/hazelcast-specific.xml]: Reverted.

**Comment by snicoll** [comment]
@leszko thank you for making your first contribution to Spring Boot and following up quickly on feedback. This is now merged in `master` with a [polish commit](https://github.com/spring-projects/spring-boot/commit/704da1750b62e192fa0ddb33ac831da40760002e). What I've changed in particular is the check on xml vs. yaml, with an explicit check on `yaml` and a fallback on xml rather than the other way around.

Also, tests assertions were all identical so that didn't really test that the right file was loaded. I couldn't fix it for system properties test of the server as I believe I've found an inconsistency in Hazelcast. I've raised https://github.com/hazelcast/hazelcast/issues/14960.

Thanks again.

**Comment by leszko** [comment]
Thanks @snicoll
